mango-core/src/main/java/org/calrissian/mango/domain/BaseAttributeStore.java

```diff
@@ -75,8 +74,19 @@ public void putAll(Iterable<Attribute> attributes) {
     }
 
     @Override
+    public <T> Attribute<T> get(String key, Class<T> clazz) {
+        Collection<Attribute> attrs = getAll(key);
+        for(Attribute attr: attrs) {
+            if(attr.getValue().getClass().isAssignableFrom(clazz))
+                return attr;
+        }
+
+        return null;
+    }
+
+    @Override
     public Set<String> keys() {
-        return attributes.keySet();
+        return new HashSet(attributes.keySet());
```

Comment:
I can understand that.  I don't understand needing to copy twice however.  Collections.unmodifiableSet() simply wraps the set and doesn't make a copy of it.  This way you don't have to make a copy unless you actually need to, which you should be doing anyway if you need to add/remove values as you don't always know if the implementation has guarded their internals correctly by either making a copy or making it unmodifiable.